### PR TITLE
Support `or` in `isinstance` checks

### LINKED ISSUE #942: Support `or` in `isinstance` expressions.
I saw in some place in mypy something like

```
class A:
    pass

class B(A):
    pass

class C(A):
    pass

x = C()  # type: A
if isinstance(x, B) or isinstance(x, C):
    y = cast(Union[B, C], x)
    ...
```

We should be able to infer `Union[B, C]` as the type of `x` within this expression.

### PR DESCRIPTION
**Pull request by rwbarton**
Fixes #942.

### CODE REVIEW

File: test-data/unit/check-isinstance.test [hunk 1]
```diff
@@ -798,6 +800,25 @@ def foo(): pass
...
+        return None
+    return x
+def h(a: bool, x: object) -> Optional[int]:
+    if a or isinstance(x, int):
+        return None
+    return x
+[builtins fixtures/isinstance.py]
+[out]
+main: note: In function "h":
+main:13: error: Incompatible return value type (got "object", expected "int")
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: Can you use a `# E:` comment here to make the test case easier to read?

**Comment by rwbarton** [comment]
Done (and rebased).

File: test-data/unit/check-isinstance.test [hunk 2]
```diff
@@ -852,3 +872,23 @@ def f(x: Union[List[int], str]) -> None:
...
+
+[case testIsinstanceOrIsinstance]
+class A: pass
+class B(A):
+    flag = 1
+class C(A):
+    flag = 2
+def f(x: A) -> int:
+    if isinstance(x, B) or isinstance(x, C):
+        return x.flag
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: I suggest revealing the type of `x` here.

File: test-data/unit/check-isinstance.test [hunk 3]
```diff
@@ -852,3 +872,23 @@ def f(x: Union[List[int], str]) -> None:
...
+    else:
+        return 0
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: Maybe also reveal the type of `x` here.

File: test-data/unit/check-isinstance.test [hunk 4]
```diff
@@ -852,3 +872,23 @@ def f(x: Union[List[int], str]) -> None:
...
+def g(x: A) -> int:
+    if isinstance(x, A) or isinstance(x, C):
+        return x.flag # E: "A" has no attribute "flag"
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: Again, reveal the type of `x` here.

File: test-data/unit/check-isinstance.test [hunk 5]
```diff
@@ -798,6 +800,24 @@ def foo(): pass
 [out]
 main: note: In function "f":
 
+[case testIsinstanceOr]
+from typing import Optional
+def f(a: bool, x: object) -> Optional[int]:
+    if a or not isinstance(x, int):
+        return None
+    return x
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: Reveal the type of `x` here.

File: test-data/unit/check-isinstance.test [hunk 6]
```diff
@@ -798,6 +800,24 @@ def foo(): pass
...
+def g(a: bool, x: object) -> Optional[int]:
+    if not isinstance(x, int) or a:
+        return None
+    return x
```
Comment (JukkaL) [inline test-data/unit/check-isinstance.test]: Reveal the type of `x` here.

File: mypy/checker.py [hunk 7]
```diff
@@ -2424,6 +2427,31 @@ def is_literal_none(n: Node) -> bool:
     return isinstance(n, NameExpr) and n.fullname == 'builtins.None'
 
 
+def and_conditional_maps(m1: Optional[TypeMap], m2: Optional[TypeMap]) -> Optional[TypeMap]:
+    if m1 is None or m2 is None:
+        # One of the branches can never occur.
+        return None
+    # Both branches are possible; combine the information.
```
Comment (JukkaL) [inline mypy/checker.py]: Add a mention that the `m2` argument takes precedence if the same node happens to be in both maps. I wonder if a more correct behavior would be to generate an intersection type? Would as we don't have intersection types, the current behavior is fine (and it's the same as it used to be).

File: mypy/checker.py [hunk 8]
```diff
@@ -2424,6 +2427,31 @@ def is_literal_none(n: Node) -> bool:
     return isinstance(n, NameExpr) and n.fullname == 'builtins.None'
 
 
+def and_conditional_maps(m1: Optional[TypeMap], m2: Optional[TypeMap]) -> Optional[TypeMap]:
```
Comment (JukkaL) [inline mypy/checker.py]: Add docstring with some motivation/example.

File: mypy/checker.py [hunk 9]
```diff
@@ -2424,6 +2427,31 @@ def is_literal_none(n: Node) -> bool:
...
+    if m1 is None or m2 is None:
+        # One of the branches can never occur.
+        return None
+    # Both branches are possible; combine the information.
+    result = m1.copy()
+    result.update(m2)
+    return result
+
+
+def or_conditional_maps(m1: Optional[TypeMap], m2: Optional[TypeMap]) -> Optional[TypeMap]:
```
Comment (JukkaL) [inline mypy/checker.py]: Add docstring with motivation/example.

File: mypy/checker.py [hunk 10]
```diff
@@ -2471,7 +2499,22 @@ def find_isinstance_check(node: Node,
...
+            weak,
+        )
+
+        right_if_vars, right_else_vars = find_isinstance_check(
+            node.right,
+            type_map,
+            weak,
+        )
+
+        return (and_conditional_maps(left_if_vars, right_if_vars),
```
Comment (JukkaL) [inline mypy/checker.py]: Could you try to describe this as it's a little tricky, or maybe give an example (doesn't need to cover all special cases, but just the general idea).

File: mypy/checker.py [hunk 11]
```diff
@@ -2482,16 +2525,9 @@ def find_isinstance_check(node: Node,
...
-            if right_if_vars is not None:
-                left_if_vars.update(right_if_vars)
-            else:
-                left_if_vars = None
-        else:
-            left_if_vars = right_if_vars
 
-        # Make no claim about the types in else
-        return left_if_vars, {}
+        return (or_conditional_maps(left_if_vars, right_if_vars),
```
Comment (JukkaL) [inline mypy/checker.py]: Again, it would be good to describe this. Perhaps just refer to the 'and' case and discuss how this is different.

**Comment by JukkaL** [comment]
LGTM! Some minor nits about test cases and comments above.

**Comment by rwbarton** [comment]
Thanks for the thorough review. I had meant to add more documentation about `TypeMap` actually, and writing these made me notice that the behavior of `isinstance(x, T) and isinstance(x, U)` was actually nondeterministic. Pushed a new version addressing this and your other comments.

**Pull request merged**